### PR TITLE
protect against process factories that are unavailable [GEOT-5418]

### PR DESCRIPTION
**Pull request by jodygarnett**
Also protect against parameters that are not available on the classpath.

 https://osgeo-org.atlassian.net/browse/GEOT-5418

### CODE REVIEW

**Comment by jodygarnett** [comment]
@tbarsballe and @jnh5y will this solve your geomesa deployment issue

**Comment by tbarsballe** [comment]
Have a bug report: https://osgeo-org.atlassian.net/browse/GEOT-5418

**Comment by tbarsballe** [comment]
This looks to have fixed the issue; GeoMesa builds after applying this change to GeoTools.
I was not able to determine what process was actually causing the error - it looks like the GeoMesa build might be suppressing some logging.

**Comment by jnh5y** [comment]
@jodygarnett @tbarsballe thanks guys!  Can we get this ported to the 15.x series as well?

**Comment by tbarsballe** [comment]
That's the plan. Ideally this will be included in the 15.0 release.
@jodygarnett - I think this is good to merge; If you want to merge to master, I can backport to 15.x.

**Comment by jnh5y** [comment]
As a note, I very, very briefly tested this change as a backport to 15-RC1, and it worked.

**Comment by simboss** [comment]
What about adding a test?

**Comment by aaime** [comment]
Agreed about the test. Should be easy and clean, the Processors class has registerProcessFactory/removeProcessFactory methods that allow the test to temporarily add a mock factory claiming not to be available.

**Comment by jodygarnett** [comment]
I have had a go at the following:
````
        ProcessFactory good = new TestFactory(new NameImpl("test", "good"), true);
        ProcessFactory evil = new TestFactory(new NameImpl("test", "evil"), false);
        ProcessFunctionFactory factory = new ProcessFunctionFactory();
        try {
            Processors.addProcessFactory(good);
            Processors.addProcessFactory(evil);
            
            List<FunctionName> list = factory.getFunctionNames();
            
            assertTrue("available", list.contains(new NameImpl("test", "good")));
            assertFalse("unavailable", list.contains(new NameImpl("test", "evil")));
        } finally {
            Processors.removeProcessFactory(good);
            Processors.removeProcessFactory(evil);
        }
````

Unfortunately the add/remove process factory methods are not being respected, the factories are not returned by `Processors.getProcessFactories()` which returns a single `BufferFeatureCollectionFactory` factory.

**Comment by aaime** [comment]
Thanks for adding the test. Not sure about the programmatic registration issue, GeoServer is using it without problems as far as I know.

File: modules/unsupported/process/src/main/java/org/geotools/process/function/ProcessFunctionFactory.java [hunk 1]
```diff
@@ -137,32 +135,37 @@ private synchronized void init() {
             processToFunction = new HashMap<Name,FunctionName>();
             functionNames = new ArrayList<FunctionName>();
             for (ProcessFactory factory : factories) {
+                if( !factory.isAvailable() ) continue;
```
Comment (aaime) [inline modules/unsupported/process/src/main/java/org/geotools/process/function/ProcessFunctionFactory.java]: Not a stopped, you might just ignore the comment, but this does not align with our coding conventions, it should have been:

````
if(!factory.isAvailable()) {
  continue;
}

File: modules/unsupported/process/src/main/java/org/geotools/process/function/ProcessFunctionFactory.java [hunk 2]
```diff
@@ -137,32 +135,37 @@ private synchronized void init() {
...
+                            FunctionName functionName = new FunctionNameImpl(processName, result, args);
+                            functionNames.add(functionName);
+                            processToFunction.put(processName, functionName);
                         }
-
-                        FunctionName functionName = new FunctionNameImpl(processName, result, args);
-                        functionNames.add(functionName);
-                        processToFunction.put(processName, functionName);
+                    }
+                    catch (Throwable unavailable){
```
Comment (aaime) [inline modules/unsupported/process/src/main/java/org/geotools/process/function/ProcessFunctionFactory.java]: Similarly here, there should have been a space between ) and {

File: modules/unsupported/process/src/test/java/org/geotools/process/factory/VectorIdentityRTProcess.java [hunk 3]
```diff
@@ -1,3 +1,19 @@
+/*
```
Comment (aaime) [inline modules/unsupported/process/src/test/java/org/geotools/process/factory/VectorIdentityRTProcess.java]: Thanks for adding the missing headers

File: modules/unsupported/process/src/main/java/org/geotools/process/function/ProcessFunctionFactory.java (hunk 2 above)
Comment (bencaradocdavies) [inline modules/unsupported/process/src/main/java/org/geotools/process/function/ProcessFunctionFactory.java]: And ```catch``` should be on the same line as the preceding closing brace:
```
} catch (Throwable unavailable) {

**Comment by jodygarnett** [comment]
Thanks for feedback, squashed and merged.

**Pull request closed without merging**